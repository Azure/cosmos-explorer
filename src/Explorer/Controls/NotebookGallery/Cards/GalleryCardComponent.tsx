import {
  BaseButton,
  Button,
  DocumentCard,
  DocumentCardActivity,
  DocumentCardDetails,
  DocumentCardPreview,
  DocumentCardTitle,
  Icon,
  IconButton,
  IDocumentCardPreviewProps,
  IDocumentCardStyles,
  ImageFit,
  Link,
  Separator,
  Spinner,
  SpinnerSize,
  Text,
  TooltipHost,
} from "@fluentui/react";
import React, { FunctionComponent, useState } from "react";
import CosmosDBLogo from "../../../../../images/CosmosDB-logo.svg";
import { IGalleryItem } from "../../../../Juno/JunoClient";
import * as FileSystemUtil from "../../../Notebook/FileSystemUtil";

export interface GalleryCardComponentProps {
  data: IGalleryItem;
  isFavorite: boolean;
  showDownload: boolean;
  showDelete: boolean;
  onClick: () => void;
  onTagClick: (tag: string) => void;
  onFavoriteClick: () => void;
  onUnfavoriteClick: () => void;
  onDownloadClick: () => void;
  onDeleteClick: (beforeDelete: () => void, afterDelete: () => void) => void;
}

export const GalleryCardComponent: FunctionComponent<GalleryCardComponentProps> = ({
  data,
  isFavorite,
  showDownload,
  showDelete,
  onClick,
  onTagClick,
  onFavoriteClick,
  onUnfavoriteClick,
  onDownloadClick,
  onDeleteClick,
}: GalleryCardComponentProps) => {
  const CARD_WIDTH = 256;
  const cardImageHeight = 144;
  const cardDescriptionMaxChars = 80;
  const cardItemGapSmall = 8;
  const cardDeleteSpinnerHeight = 360;
  const smallTextLineHeight = 18;

  const [isDeletingPublishedNotebook, setIsDeletingPublishedNotebook] = useState<boolean>(false);

  const cardButtonsVisible = isFavorite !== undefined || showDownload || showDelete;
  const options: Intl.DateTimeFormatOptions = {
    year: "numeric",
    month: "short",
    day: "numeric",
  };
  const dateString = new Date(data.created).toLocaleString("default", options);
  const cardTitle = FileSystemUtil.stripExtension(data.name, "ipynb");

  const renderTruncated = (text: string, totalLength: number): string => {
    let truncatedDescription = text.substr(0, totalLength);
    if (text.length > totalLength) {
      truncatedDescription = `${truncatedDescription} ...`;
    }
    return truncatedDescription;
  };

  const generateIconText = (iconName: string, text: string): JSX.Element => {
    return (
      <Text variant="tiny" styles={{ root: { color: "#605E5C", paddingRight: cardItemGapSmall } }}>
        <Icon iconName={iconName} styles={{ root: { verticalAlign: "middle" } }} /> {text}
      </Text>
    );
  };

  /*
   * Fluent UI doesn't support tooltips on IconButtons out of the box. In the meantime the recommendation is
   * to do the following (from https://developer.microsoft.com/en-us/fluentui#/controls/web/button)
   */
  const generateIconButtonWithTooltip = (
    iconName: string,
    title: string,
    horizontalAlign: "right" | "left",
    activate: () => void
  ): JSX.Element => {
    return (
      <TooltipHost
        content={title}
        id={`TooltipHost-IconButton-${iconName}`}
        calloutProps={{ gapSpace: 0 }}
        styles={{ root: { display: "inline-block", float: horizontalAlign } }}
      >
        <IconButton
          iconProps={{ iconName }}
          title={title}
          ariaLabel={title}
          onClick={(event) => handlerOnClick(event, activate)}
        />
      </TooltipHost>
    );
  };

  const handlerOnClick = (
    event:
      | React.MouseEvent<HTMLElement | HTMLAnchorElement | HTMLButtonElement | MouseEvent>
      | React.MouseEvent<
          HTMLAnchorElement | HTMLButtonElement | HTMLDivElement | BaseButton | Button | HTMLSpanElement,
          MouseEvent
        >,
    activate: () => void
  ): void => {
    event.stopPropagation();
    event.preventDefault();
    activate();
  };
  const DocumentCardActivityPeople = [{ name: data.author, profileImageSrc: data.isSample && CosmosDBLogo }];
  const previewProps: IDocumentCardPreviewProps = {
    previewImages: [
      {
        previewImageSrc: data.thumbnailUrl,
        imageFit: ImageFit.cover,
        width: CARD_WIDTH,
        height: cardImageHeight,
      },
    ],
  };
  const cardStyles: IDocumentCardStyles = {
    root: { display: "inline-block", marginRight: 20, width: CARD_WIDTH },
  };
  return (
    <DocumentCard aria-label={cardTitle} styles={cardStyles} onClick={onClick}>
      {isDeletingPublishedNotebook && (
        <Spinner
          size={SpinnerSize.large}
          label={`Deleting '${cardTitle}'`}
          styles={{ root: { height: cardDeleteSpinnerHeight } }}
        />
      )}
      {!isDeletingPublishedNotebook && (
        <>
          <DocumentCardActivity activity={dateString} people={DocumentCardActivityPeople} />
          <DocumentCardPreview {...previewProps} />
          <DocumentCardDetails>
            <Text variant="small" nowrap styles={{ root: { height: smallTextLineHeight, padding: "2px 16px" } }}>
              {data.tags ? (
                data.tags.map((tag, index, array) => (
                  <span key={tag}>
                    <Link onClick={(event) => handlerOnClick(event, () => onTagClick(tag))}>{tag}</Link>
                    {index === array.length - 1 ? <></> : ", "}
                  </span>
                ))
              ) : (
                <br />
              )}
            </Text>
            <DocumentCardTitle title={renderTruncated(cardTitle, 20)} shouldTruncate />
            <DocumentCardTitle
              title={renderTruncated(data.description, cardDescriptionMaxChars)}
              showAsSecondaryTitle
            />
            <span style={{ padding: "8px 16px" }}>
              {data.views !== undefined && generateIconText("RedEye", data.views.toString())}
              {data.downloads !== undefined && generateIconText("Download", data.downloads.toString())}
              {data.favorites !== undefined && generateIconText("Heart", data.favorites.toString())}
            </span>
          </DocumentCardDetails>
          {cardButtonsVisible && (
            <DocumentCardDetails>
              <Separator styles={{ root: { padding: 0, height: 1 } }} />

              <span style={{ padding: "0px 16px" }}>
                {isFavorite !== undefined &&
                  generateIconButtonWithTooltip(
                    isFavorite ? "HeartFill" : "Heart",
                    isFavorite ? "Unfavorite" : "Favorite",
                    "left",
                    isFavorite ? onUnfavoriteClick : onFavoriteClick
                  )}

                {showDownload && generateIconButtonWithTooltip("Download", "Download", "left", onDownloadClick)}

                {showDelete &&
                  generateIconButtonWithTooltip("Delete", "Remove", "right", () =>
                    onDeleteClick(
                      () => setIsDeletingPublishedNotebook(true),
                      () => setIsDeletingPublishedNotebook(false)
                    )
                  )}
              </span>
            </DocumentCardDetails>
          )}
        </>
      )}
    </DocumentCard>
  );
};
