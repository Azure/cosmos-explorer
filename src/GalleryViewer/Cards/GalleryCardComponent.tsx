import * as React from "react";
import * as DataModels from "../../Contracts/DataModels";
import { Card, ICardTokens, ICardSectionTokens } from "@uifabric/react-cards";
import { Icon, Image, Persona, Text } from "office-ui-fabric-react";
import {
  siteTextStyles,
  descriptionTextStyles,
  helpfulTextStyles,
  subtleHelpfulTextStyles,
  subtleIconStyles
} from "./CardStyleConstants";

interface GalleryCardComponentProps {
  name: string;
  url: string;
  notebookMetadata: DataModels.NotebookMetadata;
  onClick: () => void;
}

export class GalleryCardComponent extends React.Component<GalleryCardComponentProps> {
  private cardTokens: ICardTokens = { childrenMargin: 12 };
  private attendantsCardSectionTokens: ICardSectionTokens = { childrenGap: 6 };

  public render(): JSX.Element {
    return this.props.notebookMetadata != null ? (
      <Card aria-label="Notebook Card" onClick={this.props.onClick} tokens={this.cardTokens}>
        <Card.Item>
          <Persona text={this.props.notebookMetadata.author} secondaryText={this.props.notebookMetadata.date} />
        </Card.Item>
        <Card.Item fill>
          <Image src={this.props.notebookMetadata.imageUrl} width="100%" alt="Notebook display image" />
        </Card.Item>
        <Card.Section>
          <Text variant="small" styles={siteTextStyles}>
            {this.props.notebookMetadata.tags.join(", ")}
          </Text>
          <Text styles={descriptionTextStyles}>{this.props.name}</Text>
          <Text variant="small" styles={helpfulTextStyles}>
            {this.props.notebookMetadata.description}
          </Text>
        </Card.Section>
        <Card.Section horizontal tokens={this.attendantsCardSectionTokens}>
          <Icon iconName="RedEye" styles={subtleIconStyles} />
          <Text variant="small" styles={subtleHelpfulTextStyles}>
            {this.props.notebookMetadata.views}
          </Text>
          <Icon iconName="Download" styles={subtleIconStyles} />
          <Text variant="small" styles={subtleHelpfulTextStyles}>
            {this.props.notebookMetadata.downloads}
          </Text>
          <Icon iconName="Heart" styles={subtleIconStyles} />
          <Text variant="small" styles={subtleHelpfulTextStyles}>
            {this.props.notebookMetadata.likes}
          </Text>
        </Card.Section>
      </Card>
    ) : (
      <Card aria-label="Notebook Card" onClick={this.props.onClick} tokens={this.cardTokens}>
        <Card.Section>
          <Text styles={descriptionTextStyles}>{this.props.name}</Text>
        </Card.Section>
      </Card>
    );
  }
}
