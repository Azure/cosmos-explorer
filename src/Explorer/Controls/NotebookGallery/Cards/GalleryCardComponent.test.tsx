import { shallow } from "enzyme";
import React from "react";
import { GalleryCardComponent, GalleryCardComponentProps } from "./GalleryCardComponent";

describe("GalleryCardComponent", () => {
  it("renders", () => {
    const props: GalleryCardComponentProps = {
      data: {
        id: "id",
        name: "name",
        description: "description",
        author: "author",
        thumbnailUrl: "thumbnailUrl",
        created: "created",
        gitSha: "gitSha",
        tags: ["tag"],
        isSample: false,
        downloads: 0,
        favorites: 0,
        views: 0,
        newCellId: undefined,
        policyViolations: undefined,
        pendingScanJobIds: undefined,
      },
      isFavorite: false,
      showDownload: true,
      showDelete: true,
      onClick: undefined,
      onTagClick: undefined,
      onFavoriteClick: undefined,
      onUnfavoriteClick: undefined,
      onDownloadClick: undefined,
      onDeleteClick: undefined,
    };

    const wrapper = shallow(<GalleryCardComponent {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
