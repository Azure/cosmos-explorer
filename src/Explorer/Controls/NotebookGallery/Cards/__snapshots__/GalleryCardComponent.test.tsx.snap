// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryCardComponent renders 1`] = `
<StyledDocumentCardBase
  aria-label="name"
  styles={
    Object {
      "root": Object {
        "display": "inline-block",
        "marginRight": 20,
        "width": 256,
      },
    }
  }
>
  <StyledDocumentCardActivityBase
    activity="Invalid Date"
    people={
      Array [
        Object {
          "name": "author",
          "profileImageSrc": false,
        },
      ]
    }
  />
  <StyledDocumentCardPreviewBase
    previewImages={
      Array [
        Object {
          "height": 144,
          "imageFit": 2,
          "previewImageSrc": "thumbnailUrl",
          "width": 256,
        },
      ]
    }
  />
  <StyledDocumentCardDetailsBase>
    <Text
      nowrap={true}
      styles={
        Object {
          "root": Object {
            "height": 18,
            "padding": "2px 16px",
          },
        }
      }
      variant="small"
    >
      <span
        key="tag"
      >
        <StyledLinkBase
          onClick={[Function]}
        >
          tag
        </StyledLinkBase>
      </span>
    </Text>
    <StyledDocumentCardTitleBase
      shouldTruncate={true}
      title="name"
    />
    <StyledDocumentCardTitleBase
      showAsSecondaryTitle={true}
      title="description"
    />
    <span
      style={
        Object {
          "padding": "8px 16px",
        }
      }
    >
      <Text
        styles={
          Object {
            "root": Object {
              "color": "#605E5C",
              "paddingRight": 8,
            },
          }
        }
        variant="tiny"
      >
        <Icon
          iconName="RedEye"
          styles={
            Object {
              "root": Object {
                "verticalAlign": "middle",
              },
            }
          }
        />
         
        0
      </Text>
      <Text
        styles={
          Object {
            "root": Object {
              "color": "#605E5C",
              "paddingRight": 8,
            },
          }
        }
        variant="tiny"
      >
        <Icon
          iconName="Download"
          styles={
            Object {
              "root": Object {
                "verticalAlign": "middle",
              },
            }
          }
        />
         
        0
      </Text>
      <Text
        styles={
          Object {
            "root": Object {
              "color": "#605E5C",
              "paddingRight": 8,
            },
          }
        }
        variant="tiny"
      >
        <Icon
          iconName="Heart"
          styles={
            Object {
              "root": Object {
                "verticalAlign": "middle",
              },
            }
          }
        />
         
        0
      </Text>
    </span>
  </StyledDocumentCardDetailsBase>
  <StyledDocumentCardDetailsBase>
    <Separator
      styles={
        Object {
          "root": Object {
            "height": 1,
            "padding": 0,
          },
        }
      }
    />
    <span
      style={
        Object {
          "padding": "0px 16px",
        }
      }
    >
      <StyledTooltipHostBase
        calloutProps={
          Object {
            "gapSpace": 0,
          }
        }
        content="Favorite"
        id="TooltipHost-IconButton-Heart"
        styles={
          Object {
            "root": Object {
              "display": "inline-block",
              "float": "left",
            },
          }
        }
      >
        <CustomizedIconButton
          ariaLabel="Favorite"
          iconProps={
            Object {
              "iconName": "Heart",
            }
          }
          onClick={[Function]}
          title="Favorite"
        />
      </StyledTooltipHostBase>
      <StyledTooltipHostBase
        calloutProps={
          Object {
            "gapSpace": 0,
          }
        }
        content="Download"
        id="TooltipHost-IconButton-Download"
        styles={
          Object {
            "root": Object {
              "display": "inline-block",
              "float": "left",
            },
          }
        }
      >
        <CustomizedIconButton
          ariaLabel="Download"
          iconProps={
            Object {
              "iconName": "Download",
            }
          }
          onClick={[Function]}
          title="Download"
        />
      </StyledTooltipHostBase>
      <StyledTooltipHostBase
        calloutProps={
          Object {
            "gapSpace": 0,
          }
        }
        content="Remove"
        id="TooltipHost-IconButton-Delete"
        styles={
          Object {
            "root": Object {
              "display": "inline-block",
              "float": "right",
            },
          }
        }
      >
        <CustomizedIconButton
          ariaLabel="Remove"
          iconProps={
            Object {
              "iconName": "Delete",
            }
          }
          onClick={[Function]}
          title="Remove"
        />
      </StyledTooltipHostBase>
    </span>
  </StyledDocumentCardDetailsBase>
</StyledDocumentCardBase>
`;
